### PR TITLE
Use Scala 3

### PR DESCRIPTION
**Pull request by kubukoz**
circe-generic-extras being missing is probably the biggest pain point, using a custom solution for now.

### CODE REVIEW

**Comment by kubukoz** [comment]
@majk-p we'll finally be able to reuse some gitlab code in the bootstrap 😅

File: src/test/scala/io/pg/ProjectConfigFormatTests.scala [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+import io.pg.gitlab.webhook.Project
+import io.pg.config.Rule
+import io.pg.config.Action
+import io.pg.config.Matcher
+import io.pg.config.TextMatcher
+import io.circe.syntax._
+
+object ProjectConfigFormatTest extends FunSuite {
+
+  val asJSON = json"""{
```
Comment (kubukoz) [inline src/test/scala/io/pg/ProjectConfigFormatTests.scala:16]: Generated based on a dump from `dhall-to-json`

File: src/main/scala/io/pg/transport/transport.scala [hunk 2]
```diff
@@ -1,43 +1,25 @@
 package io.pg.transport
 
-import io.circe.generic.extras.Configuration
-import io.circe.generic.extras.ConfiguredJsonCodec
-import io.circe.generic.extras.semiauto.deriveEnumerationCodec
 import io.circe.Codec
 
-private object CirceConfiguration {
-  implicit val circeConfig: Configuration = Configuration.default.withDiscriminator("@type")
-}
```
Comment (kubukoz) [inline src/main/scala/io/pg/transport/transport.scala:10]: Technically this changes the format, but this was meant for debugging only anyway.

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala [hunk 3]
```diff
@@ -62,20 +60,19 @@ object Gitlab {
...
 
   object MergeRequestInfo {
-    sealed trait Status extends Product with Serializable
 
-    object Status {
-      case object Success extends Status
-      final case class Other(value: String) extends Status
+    enum Status {
+      case Success
+      case Other(value: String)
```
Comment (majk-p) [inline gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:66]: :heart:

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala [hunk 4]
```diff
@@ -278,20 +275,19 @@ object GitlabEndpoints {
...
     object ApprovalRule {
-      implicit val codec: CirceCodec[ApprovalRule] = deriveConfiguredCodec
+      given CirceCodec[ApprovalRule] = CirceCodec.forProduct3("id", "name", "rule_type")(apply)(r => (r.id, r.name, r.ruleType))
     }
 
     final case class MergeRequestApprovals(approvalsRequired: Int)
 
     object MergeRequestApprovals {
-      implicit val codec: CirceCodec[MergeRequestApprovals] = deriveConfiguredCodec
+      given CirceCodec[MergeRequestApprovals] = CirceCodec.forProduct1("approvals_required")(apply)(_.approvalsRequired)
```
Comment (majk-p) [inline gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:290]: How about some todos to get back to `circe-generic-extras` when possible?

File: src/main/scala/io/pg/webhook/webhook.scala [hunk 5]
```diff
@@ -31,7 +33,28 @@ object WebhookRouter {
...
+                case MergeRequestInfo.Status.Other(s) => transport.MergeRequestState.Status.Other(s)
+              },
+              mergeability = s.mergeability match {
+                case Mergeability.CanMerge     => transport.MergeRequestState.Mergeability.CanMerge
+                case Mergeability.HasConflicts => transport.MergeRequestState.Mergeability.HasConflicts
+                case Mergeability.NeedsRebase  => transport.MergeRequestState.Mergeability.NeedsRebase
+              },
+              authorUsername = s.authorUsername
+            )
+          }
```
Comment (majk-p) [inline src/main/scala/io/pg/webhook/webhook.scala:55]: Nitpick: extract the transformer

**Review by majk-p — APPROVED** [review]
:+1: few minor things and LGMT

File: gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala (hunk 4 above)
Comment (kubukoz) [reply gitlab/src/main/scala/io/pg/gitlab/Gitlab.scala:290]: Might take a while 😬 I'll add some

File: core/src/main/scala/io/pg/messaging/messaging.scala [hunk 6]
```diff
@@ -35,12 +36,17 @@ object Processor {
 
 }
 
-@autoInvariant
 trait Channel[F[_], A] extends Publisher[F, A] { self =>
```
[the remark this answers is not in the document]
Comment (kubukoz) [reply core/src/main/scala/io/pg/messaging/messaging.scala:39]: nope

**Pull request merged**